pydantic/utils.py

```diff
@@ -787,3 +787,29 @@ def __setitem__(self, __key: Any, __value: Any) -> None:
         def __class_getitem__(cls, *args: Any) -> Any:
             # to avoid errors with 3.7
             pass
+
+
+URI_RESERVED_CHARACTERS = frozenset(":/?#[]@!$&'()*+,;=% ")
+
+
+def percent_encode(string: str, *, plus: bool = False) -> str:
```

Comment:
what's wrong with `urllib.parse.quote`?

Reply:
I just didn't feel like introducing a new import to pydantic since `urllib.prase` is a big module which can slow down importing pydantic but now after testing with `-X importtime`, it seems that there's no difference since `pathlib` already imports `urllib.parse`.

Here's some benchmarks for both functions tested on Python 3.10.5, let me know what you think

```
$ ./percent_encode_urllib_parse_benchmark 
short(16ch):
pydantic.utils.percent_encode: 50000 loops, best of 5: 4.4 usec per loop
urllib.parse.quote:            50000 loops, best of 5: 2.53 usec per loop

long(256ch):
pydantic.utils.percent_encode: 50000 loops, best of 5: 12.1 usec per loop
urllib.parse.quote:            50000 loops, best of 5: 18 usec per loop

large(4096ch):
pydantic.utils.percent_encode: 50000 loops, best of 5: 77.9 usec per loop
urllib.parse.quote:            50000 loops, best of 5: 255 usec per loop
```

```
$ ./percent_encode_urllib_parse_benchmark --plus
short(16ch):
pydantic.utils.percent_encode: 50000 loops, best of 5: 5.39 usec per loop
urllib.parse.quote:            50000 loops, best of 5: 2.59 usec per loop

long(256ch):
pydantic.utils.percent_encode: 50000 loops, best of 5: 15 usec per loop
urllib.parse.quote:            50000 loops, best of 5: 17.9 usec per loop

large(4096ch):
pydantic.utils.percent_encode: 50000 loops, best of 5: 79.5 usec per loop
urllib.parse.quote:            50000 loops, best of 5: 260 usec per loop
```

```python
#!/usr/bin/env python

import random
import string
import sys
import timeit

TIMEIT_ARGV = ["-n", "50000", "-u", "usec"]

if "--plus" not in sys.argv:
    quote_timeit_argv = TIMEIT_ARGV + ["--setup", "from urllib.parse import quote"]
    percent_encode_timeit_argv = TIMEIT_ARGV + ["--setup", "from pydantic.utils import percent_encode"]
else:
    quote_timeit_argv = TIMEIT_ARGV + ["--setup", "from urllib.parse import quote_plus as quote"]
    percent_encode_timeit_argv = TIMEIT_ARGV + [
        "--setup",
        ";".join(
            [
                "from functools import partial",
                "from pydantic.utils import percent_encode",
                "percent_encode = partial(percent_encode, plus=True)",
            ]
        ),
    ]


def benchmark(string: str) -> None:
    print("pydantic.utils.percent_encode: ", end="")
    timeit.main(percent_encode_timeit_argv + [f"percent_encode({string!r}, plus=True)"])
    print("urllib.parse.quote:            ", end="")
    timeit.main(quote_timeit_argv + [f"quote({string!r})"])


SHORT_STRING = "".join(random.choices(string.printable, k=16))
LONG_STRING = "".join(random.choices(string.printable, k=256))
LARGE_STRING = "".join(random.choices(string.printable, k=4096))

print(f"short({len(SHORT_STRING)}ch):")
benchmark(SHORT_STRING)

print(f"\nlong({len(LONG_STRING)}ch):")
benchmark(LONG_STRING)

print(f"\nlarge({len(LARGE_STRING)}ch):")
benchmark(LARGE_STRING)
```

Here's also a plot for strings with the length 1-24 to the power of 2, lower is bettter:

![percent_encode_urllib_parse_benchmark_plot](https://user-images.githubusercontent.com/73843130/183791974-6a8f6c94-4bb2-4715-935c-5a20cb19899c.png)